### PR TITLE
Fix slow cli archive for datasets

### LINKED ISSUE #1048: Slow `dataset archive` on a large list of UUIDs
### Expected behaviour

`dataset archive <bunch of uuids>` should be similar in speed to `dc.index.datasets.archive(..)` which is fast, but CLI version is slow.

### Actual behaviour

- For every input UUID
  - Lookup Dataset object
  - Do nothing with it other than to check if that UUID exists and is not archived already
  - Call `.archive(..)` on each UUID individually

https://github.com/opendatacube/datacube-core/blob/9517b1a332efdaa09c8e767aa8af23dc87ac150c/datacube/scripts/dataset.py#L472-L483

If user is not requesting to archive derived datasets then we should not even read anything, just collate a list of UUIDs from CLI and call `.archive(..)` once with that whole list.

### PR DESCRIPTION
**Pull request by Max-AR**
### Reason for this pull request

This PR should resolve https://github.com/opendatacube/datacube-core/issues/1048 by skipping heavy bootstrapping of the `DatasetResource` object and just querying for UUIDs in the database.

This should also partly solve;

https://github.com/opendatacube/datacube-core/issues/1029

I've not been able to test this completely on a large dataset. 

 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by Kirill888** [comment]
@Max-AR thank you for the PR, it's great. I'd like to request some changes though :). I would like to avoid making changes to the DB driver API, even though we only have one implementation of it.

We already have `index.dataset.bulk_has` method defined

https://github.com/opendatacube/datacube-core/blob/afca22831eecf8092e0c293064a9ea465aa6d778/datacube/index/_datasets.py#L137-L146
https://datacube-core.readthedocs.io/en/latest/dev/api/generate/datacube.index._datasets.DatasetResource.bulk_has.html

Can I ask you to revert `exists` and `exists_dataset` methods and use `bulk_has` in the archival code instead?

**Comment by Max-AR** [comment]
@Kirill888 Thanks for looking over this PR :)  

I've made those changes you were after with the DB API. I double-checked the tests and it seems we were not catching all the cases that were needed for proper coverage, so I have expanded test cases as well. 

I do have two small reservations about the PR as it stands:

- The error code for `clirunner` is always checked for `sys.exit(0)` in the tests, so I've updated the "UUID not found" case to not exit with an error code. It would be a bit nicer for anyone scripting with the `datacube` cli to get those exit codes into bash.
- In my (small) testing, the DB always comes back with the `bulk_has` list in order, though for larger sets I am not 100% sure that order will be guaranteed, causing the `zip` in that dictionary comprehension to cause errors further down.

At any rate, let me know what you think!

**Comment by Kirill888** [comment]
`bulk_has` output order is guaranteed to be in the input order, it is set in Python not by DB. DB returns a subset of UUIDs that are present in the database, and then python uses that to compute present/not present flag.

File: datacube/scripts/dataset.py [hunk 1]
```diff
@@ -469,18 +469,29 @@ def _get_derived_set(index: Index, id_: str) -> Set[Dataset]:
...
-        for d in to_process:
-            click.echo('archiving %s %s %s' % (d.type.name, d.id,
-                                               d.local_uri if d.local_uri is not None else "<no location>"))
-        if not dry_run:
-            index.datasets.archive(d.id for d in to_process)
+    if False in datasets_for_archive.values():
+        for dataset_id, exists in datasets_for_archive.items():
+            if not exists:
+                click.echo(f'No dataset found with id: {dataset_id}')
+        sys.exit(0)
```
Comment (Kirill888) [inline datacube/scripts/dataset.py]: This should not be `0`.

You can use `expect_success=False` argument when using `clirunner`, it won't assert then. You can then check exit code is of expected value in the test.

Comment (Kirill888) [reply datacube/scripts/dataset.py]: Example of checking for exit code other than 0

https://github.com/opendatacube/datacube-core/blob/afca22831eecf8092e0c293064a9ea465aa6d778/integration_tests/test_validate_ingestion.py#L31-L35

Comment (Max-AR) [reply datacube/scripts/dataset.py]: Thanks for pointing me in the right direction, I've updated the code :+1:

**Review by Kirill888 — APPROVED** [review]
thanks for making all these changes

**Pull request merged**